### ISSUE TITLE
GradientCheckpointing<T>.Checkpoint backward broadcast fails on Transformer<T>: [1,256] vs [1,seqLen,dModel]

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#1341]
## Summary

Calling `network.EnableMemoryManagement(TrainingMemoryConfig.ForTransformers())` on a `Transformer<float>` and then `Train`-ing per-sample throws `ArgumentException` in `GradientCheckpointing<T>.Checkpoint`'s backward closure:

```
Tensors with shapes [1, 256] and [1, 64, 128] cannot be broadcast: dimension 2 has sizes 256 and 128 (must be equal or one must be 1).
```

The first shape is the loss/target tensor `[batch=1, vocab=256]`, the second is the encoder activation `[batch=1, ctxLen=64, dModel=128]`.

## Stack trace

```
at AiDotNet.Tensors.LinearAlgebra.Tensor`1.GetBroadcastShape(Int32[] shape1, Int32[] shape2)
at AiDotNet.Tensors.LinearAlgebra.Tensor`1.BroadcastMultiply(Tensor`1 other)
at AiDotNet.Tensors.Engines.CpuEngine.TensorBroadcastMultiply[T](Tensor`1 a, Tensor`1 b)
at AiDotNet.Tensors.Engines.DirectGpuTensorEngine.TensorBroadcastMultiply[T](Tensor`1 a, Tensor`1 b)
at AiDotNet.Tensors.Engines.CpuEngine.TensorMultiply[T](Tensor`1 a, Tensor`1 b)
at AiDotNet.Tensors.Engines.Autodiff.GradientCheckpointing\`1.<>c__DisplayClass0_0.<Checkpoint>b__0(Tensor`1 gradOutput, Tensor`1[] inputs, Tensor`1 output, Object[] savedState, IEngine eng, Dictionary`2 grads)
at AiDotNet.Tensors.Engines.Autodiff.CompiledDelegateChain`1.Execute(...)
at AiDotNet.Tensors.Engines.Autodiff.GradientTape`1.ComputeGradientsViaGraphCore(...)
at AiDotNet.Tensors.Engines.Autodiff.GradientTape`1.ComputeGradients(...)
at AiDotNet.NeuralNetworks.NeuralNetworkBase`1.TrainWithTape(Tensor`1 input, Tensor`1 expected, IGradientBasedOptimizer`3 optimizer)
at AiDotNet.NeuralNetworks.Transformer`1.Train(Tensor`1 input, Tensor`1 expectedOutput)
```

## Repro config

| Knob | Value |
| --- | --- |
| Architecture | `TransformerArchitecture<float>(SequenceClassification)` |
| `numEncoderLayers` | 2 |
| `numHeads` | 2 |
| `modelDimension` | 128 |
| `feedForwardDimension` | 256 |
| `inputSize` / `maxSequenceLength` / `ctxLen` | 64 |
| `outputSize` / `vocabSize` | 256 |
| Train path | per-sample `model.Train(x[1,64], y[1,256])` |
| Optimizer | `AdamOptimizer<float, Tensor<float>, Tensor<float>>` |
| Memory mgmt | `model.EnableMemoryManagement(TrainingMemoryConfig.ForTransformers())` |
| AiDotNet | 0.198.0 |
| AiDotNet.Tensors | 0.81.0 |

The exact same `Transformer<float>` + optimizer + per-sample Train loop succeeds when `EnableMemoryManagement` is NOT called (verified — produces top-1 ~9% / ppl 23 on WT2 10KB byte-LM).

## Hypothesis on root cause

`GradientCheckpointing<T>.Checkpoint`'s backward closure appears to apply `TensorMultiply(gradOutput, output)` directly without honoring the per-layer chain's expected gradient shape contract. When wrapped around the full Transformer forward (encoder → pool → vocab projection → loss), the closure receives:
- `gradOutput` shape: `[batch, vocab]` (from the loss head)
- `output` shape: `[batch, seqLen, dModel]` (the checkpointed-segment output — the encoder activations before the head)

These can't be broadcast — the gradient flow needs to route through the projection layer to reshape `[batch, vocab]` → `[batch, seqLen, dModel]` before the encoder gradient can be computed. The current implementation either:
1. Wraps the entire forward (including the vocab head) as one checkpoint segment, but only recomputes the encoder portion, or
2. Wraps the encoder portion but tries to multiply by a `gradOutput` that hasn't been transformed through the vocab head's backward.

Likely fix surface: `Tensors/src/Engines/Autodiff/GradientCheckpointing.cs` (the `Checkpoint` backward lambda) needs to either (a) include the entire forward+loss as one checkpoint scope, or (b) only checkpoint per-layer segments that share input/output shape contracts.

## Impact

Blocks consumer-side use of `EnableMemoryManagement` for memory-efficient Transformer training. Currently HarmonicEngine PAPER-A Path B can't scale to `dModel >= 256 / layers >= 4` because the alternative memory-fix path (per-call `Train` leak ~1.1 MB/call) is being addressed separately but doesn't shrink the per-step activation memory. Gradient checkpointing was the intended workaround.

## Minimal repro

See \`HarmonicEngine/tests/Learning/Phase_PAPER_A_PathB_GradientCheckpointing_Sanity.cs\` on branch \`research/ledger-overhaul-and-l4-reification\` for the failing test. The non-checkpointed arm passes; the checkpointed arm throws the broadcast exception in `Transformer<float>.Train` on the first sample.
